boilerplate/app/components/Icon.tsx

```diff
@@ -0,0 +1,99 @@
+import * as React from "react"
+import { ComponentType } from "react"
+import {
+  Image,
+  ImageStyle,
+  StyleProp,
+  TouchableOpacity,
+  TouchableOpacityProps,
+  View,
+  ViewStyle,
+} from "react-native"
+
+export type IconTypes = keyof typeof iconRegistry
+
+interface IconProps extends TouchableOpacityProps {
+  /**
+   * The name of the icon
+   */
+  icon: IconTypes
+
+  /**
+   * An optional tint color for the icon
+   */
+  color?: string
+
+  /**
+   * An optional size for the icon. If not provided, the icon will be sized to the icon's resolution.
+   */
+  size?: number
+
+  /**
+   * Style overrides for the icon image
+   */
+  style?: StyleProp<ImageStyle>
+
+  /**
+   * Style overrides for the icon container
+   */
+  containerStyle?: StyleProp<ViewStyle>
+}
+
+/**
+ * A component to render a registered icon.
+ * It is wrapped in a <TouchableOpacity /> if `onPress` is provided, otherwise a <View />.
+ *
+ * - [Documentation and Examples](https://github.com/infinitered/ignite/blob/master/docs/Components-Icon.md)
+ */
+export function Icon(props: IconProps) {
+  const {
+    icon,
+    color,
+    size,
+    style: $imageStyleOverride,
+    containerStyle: $containerStyleOverride,
+    onPress,
+    ...WrapperProps
+  } = props
+
+  const Wrapper: ComponentType<TouchableOpacityProps> = onPress ? TouchableOpacity : View
+
+  return (
+    <Wrapper {...WrapperProps} style={$containerStyleOverride}>
+      <Image
+        style={[
+          $imageStyle,
+          { tintColor: color },
+          size && { width: size, height: size },
+          $imageStyleOverride,
+        ]}
+        source={iconRegistry[icon]}
+      />
+    </Wrapper>
+  )
+}
+
+export const iconRegistry = {
+  back: require("../../assets/icons/back.png"),
+  bell: require("../../assets/icons/bell.png"),
+  caretLeft: require("../../assets/icons/caretLeft.png"),
+  caretRight: require("../../assets/icons/caretRight.png"),
+  check: require("../../assets/icons/check.png"),
+  community: require("../../assets/icons/community.png"),
+  components: require("../../assets/icons/components.png"),
+  debug: require("../../assets/icons/debug.png"),
+  heart: require("../../assets/icons/heart.png"),
+  hidden: require("../../assets/icons/hidden.png"),
+  ladybug: require("../../assets/icons/ladybug.png"),
+  lock: require("../../assets/icons/lock.png"),
+  menu: require("../../assets/icons/menu.png"),
+  more: require("../../assets/icons/more.png"),
+  pin: require("../../assets/icons/pin.png"),
+  settings: require("../../assets/icons/settings.png"),
+  view: require("../../assets/icons/view.png"),
+  x: require("../../assets/icons/x.png"),
+}
```

Comment:
hmmm, can this somehow pull automagically? would be tough to always have to update this list when adding a new icon

Reply:
Good question. I'm not sure entirely how to do this dynamically (including getting the TS types for the icon prop). Also, I think if you do a dynamic require, it breaks metro. Let me know if you have ideas.

Reply:
Might be a good use case for a generator that works like https://github.com/elderfo/react-native-storybook-loader creating this registry with a cli command.

Reply:
Generator would be cool - we size the icon correctly, move it to the correct folder, than update a registry file. Might be out of scope for now though.

Maybe what we could do meanwhile is just run a prestart script that builds the icon registry file based on the contents of the icons folder. Basically automate the above requires. The downside is that people would need to restart their metro instance if they add new icons and don't want to manually update the file. Thoughts?

Reply:
Running a script on start up would be rad. Once you have a script you could manually run via cli as well so that would cover not needing to restart metro.

Reply:
my initial thought of having an `index.ts` file that exports all the icons and then TS to handle the rest didn't pan out.

BUT your idea on a prestart script would be really cool. I vote for that!

Reply:
Good idea -- post-Maverick